### ISSUE TITLE
Could this library support customized function as comparator for priority property?

### ISSUE THREAD

**Issue by jialihan** [datastructures-js/priority-queue#20]
Hey, when I use this in leetcode, the use case is when I want a complicated comparator rather than simply compare by one value:
```
const biddersQueue = new MaxPriorityQueue({ priority: (bid) => bid.value });
// something i want:
const biddersQueue = new MaxPriorityQueue({ priority: (a,b) =>{  // comparator function} } );
```
maybe can this library support a comparator function in constructor?

**Comment by eyas-ranjous** [comment]
Hi there,

the logic in this library is building types for the priority queue, MaxPriorityQueue (like MinPriorityQueue) are the two types that define comparison internally between two elements in the queue (which is an instance of a Min/Max Heap). priority should is a defined numeric value for each element that is being added. because it will be the element's key in the Heap. So by design, no complex comparison should be needed because in case of Max, the check is `parentKey > childKey`, and in case of Min, the check `parentKey < childKey`. and as keys are numbers, that fulfills the check.

I'm not sure why you need a comparator function when using this design here. the idea of the callback in the constructor is to get the priority number from the queued element using any complex logic

```js
const biddersQueue = new MaxPriorityQueue({ priority: (obj) => // any complex logic to produce a priority number });
```

Let me know if that does not help. by clarifying exactly how you are using the queue in your solution.

Cheers.
Eyas.

**Comment by emreozdincer** [comment]
Hi @eyas-ranjous, I feel the same need as @jialihan. 

One example is [this question](https://leetcode.com/problems/top-k-frequent-words/). Here, we are required to sort words by frequency, but if the word frequencies of two words are the same, then we need to sort them alphabetically.

For example, this is the initialization in Java:

```java
PriorityQueue<Map.Entry<String, Integer>> pq = new PriorityQueue<>(
                 (a,b) -> a.getValue()==b.getValue() ? b.getKey().compareTo(a.getKey()) : a.getValue()-b.getValue()
        );
```

Here is a [mocked-up implementation](https://leetcode.com/problems/top-k-frequent-words/discuss/651714/javascript-structured-min-heap-(simulated-for-interview)) of the expected functionality in JS.

Basically, priority number of a single item isn't enough to decide the result of the comparison - we need to be able to write a custom comparator.

Cheers

**Comment by eyas-ranjous** [comment]
Hi @emreozdincer thanks for the examples. I will work something out to support this.

**Comment by ohpyupi** [comment]
Any plan on implementing the requested feature?

**Comment by eyas-ranjous** [comment]
Hi @ohpyupi. Yes, I will implement this, it needs changes in the heap that the queue here uses. I'm a little busy atm, hopefully to get this done early next month.

**Comment by eyas-ranjous** [comment]
https://github.com/datastructures-js/heap/issues/14

**Comment by eyas-ranjous** [comment]
feature released.

usage: https://github.com/datastructures-js/priority-queue#with-comparator
